src/h5cpp/node/dataset.hpp

```diff
@@ -947,7 +947,7 @@ template<typename T>
 void Dataset::read(T &data,const property::DatasetTransferList &dtpl) const
 {
   auto memory_space = hdf5::dataspace::create(data);
-  auto memory_type  = hdf5::datatype::create(data);
+  auto memory_type = datatype();
```

Comment:
The memory data type can be different than the file data type. The hdf5 library can automatically convert it (although it is time consuming). I don't think this change is OK because it may cause errors when `T` type does not match to `Dataset.datatype()`.

Reply:
I think I made this change in response to a failing test when reading a string type. As the tests have all changed I've reverted this change and rebased on the latest master.